client/service/src/client/client.rs

```diff
@@ -1699,6 +1711,15 @@ impl<B, E, Block, RA> sp_consensus::BlockImport<Block> for &Client<B, E, Block,
 		let span = tracing::span!(tracing::Level::DEBUG, "import_block");
 		let _enter = span.enter();
 
+		// Pre-announce the block since it's checked and validated.
+		// This removes the latency of having to broadcast the block
+		// to connected peers.
+		self.notify_pre_imported(import_block.header.hash())
```

Comment:
Can you please elaborate this design decision?

Why do we once drag the block through the full system, send some pre import notification and than notify the sync again?

With this change it also means that we will announce the block before importing, when we have build the block our own.

Reply:
Why isn't the sync layer in control over this, as it notifies the block import anyway?

This would mean, request a header check, the header is checked, we re-announce the block and after that we continue to import the block.

Yes this will clearly require more changes, but will be less spaghetti code.

Reply:
Sure, I suggested to do it this way.

> Why do we once drag the block through the full system, send some pre import notification and than notify the sync again?

Because that is what was asked for. Specifically, [as the ticket stated](https://github.com/paritytech/substrate/issues/7926) to announce the block (over the network) once after initial checks are done and once after the actual import happened.

The instance that currently triggers that 'announcing' is this components, doing that through the channels. To provide this features for all consensus engines, we do that once here when they trigger the actual import (after they have done the light-weight checks) but before heavy lifting of actual execution happens.

> With this change it also means that we will announce the block before importing
That is the intention, yes. Please see the ticket.

> when we have build the block our own.
You correct, I missed that in my review, even though I said it before, This need to do a similar `make_notifications` check on `BlockOrigin` as execute does again later, but also filter out `Own`:
```rust
		let make_notifications = match import_block.origin {
			BlockOrigin::NetworkBroadcast |  BlockOrigin::ConsensusBroadcast => true,
			BlockOrigin::Genesis | BlockOrigin::NetworkInitialSync | BlockOrigin::Own | BlockOrigin::File => false,
		};
```
> Why isn't the sync layer in control over this, as it notifies the block import anyway?
> 
> This would mean, request a header check, the header is checked, we re-announce the block and after that we continue to import the block.

Not quite sure, what component you exactly mean here. To avoid confusion, could you point me to the part of that you are referring to?

The reason to put this here was because this isn't only checking the general signatures, but after the consensus's own `verify` has been run. So, not only is the 'header checked' to not be incorrect, the generic consensus check also happened, giving us more confidence that we are announcing a block we are probably not going to fail to import. If we announced without the consensus checks, we'd have a higher risk of announcing bad blocks - e.g. one check makes sure the block is actually not running against a finalized branch. It's a trade-off for sure but as the consensus checks are - compared to the actual state changes of an import - still pretty light weight, this is a decent optimization to comparable little risk.

That's the reasoning behind locating the pre-announcement here, in the 'consensus importing'-area, to have the validity of the general consensus checks and lower the risk of pre-announcing invalid blocks. Right now, there is no signal or other way to inform the outside that a block has reached that point. Hence the idea to extend the existing channel with the new signal.

An alternative approach we've considered would have been to create a yet another channel for that signal, but extending the existing channel with other types of signals and thus allow others to also react - e.g. this would allow for offchain-workers to be informed about a block pre-import - seemed more sensible.

Reply:
> Not quite sure, what component you exactly mean here. To avoid confusion, could you point me to the part of code that you are referring to?

Here we are notified about the block data: https://github.com/paritytech/substrate/blob/master/client/network/src/protocol.rs#L819

Than it will be processed here:
https://github.com/paritytech/substrate/blob/master/primitives/consensus/common/src/import_queue.rs#L176

And than it will be notified here again:
https://github.com/paritytech/substrate/blob/master/client/network/src/protocol.rs#L1293

Basically there is a lot of stuff that needs to be improved, especially in `import_queue.rs`, but this should be the main interaction point. We should communicate the result from there back to the requesting side, in our case the sync, that the verification was successful. 

That the sync layer is currently informed about what to announce over the service is also some old "artifact", which is mainly required for when WE build a block, but should not be required for when the sync layer requests to import something.

We should really try to prevent adding some new notification that is added for a subsystem that already has a direct "communication channel" with the block import. 


> e.g. this would allow for offchain-workers to be informed about a block pre-import - seemed more sensible.

Not sure what the benefit of this should be. Offchain-workers are executed on the state of the imported block, aka they can not do anything with this information that a block was verified correctly.

Reply:
@bkchr As @gnunicorn mentioned, we opted to announce a pre-imported block right after the block was verified. in `import_single_block_metered` in `import_queue` is called by `import_many_blocks` after which `on_blocks_process` is called. At this point in `on_blocks_processed` all blocks have been imported which is a bit too late for the pre-import block rebroadcast to happen. We could have placed the pre-import notification in `import_single_block_metered` but @gnunicorn's opinion about this was that `Client::check_block` already contained the implementation for `finalized` and `import` notifications after executing the block so the best way for us was to announce pre-import there before the block was executed. As the discussion in the issue mentions, what we want to achieve is announcing the block **before** the block was executed as latency is incurred there.

> We should really try to prevent adding some new notification that is added for a subsystem that already has a direct "communication channel" with the block import.

I guess this is your biggest objection to the current implementation. I don't 100% understand what you mean by "direct communication channel with block import". But since `import_many_blocks` communicates `on_blocks_processed` back to the network through `BufferedLink`, we could atler the code here:
https://github.com/paritytech/substrate/blob/0b719f85222c85ff095585d128fac1979bbfdd38/client/network/src/service.rs#L1819-L1834
To add an additional method to the trait `on_block_verified` to the trait & the `impl` blocks so that this method once called, would handle the announcement of pre-imports... i am not 100% that would work but i am trying to find a middle ground here.
What do you think?

Reply:
> @bkchr As @gnunicorn mentioned, we opted to announce a pre-imported block right after the block was verified. in `import_single_block_metered` in `import_queue` is called by `import_many_blocks` after which `on_blocks_process` is called.

You know that you always only import one block when you sync at the tip of the chain? Maybe there are 2 when there is a fork, but it is very very unlikely that you download both forks at the same time.

> As the discussion in the issue mentions, what we want to achieve is announcing the block **before** the block was executed as latency is incurred there.

I can also repeat myself again, I have read the issue and understand the issue ;)



> I don't 100% understand what you mean by "direct communication channel with block import".

There is already a "channel" between the sync layer and the import queue. As already explained somewhere here, the following happens: Sync downloads a block, issues a block import, the block is imported and sync is informed about the imported block. What you want, is a step in between, that sends a notification back on the same "channel" as "block is imported".

The main problem here is that the current interfaces and implementation is shit in Substrate. We have this weird `NetworkLink` etc stuff that communicates with the `import-queue`.  

We could probably go the middle way you have proposed with the `on_block_verified` method. This is still better than adding a new announcement that is wired through the whole code base :+1: